controller/orphan_controller.go

```diff
@@ -269,48 +303,200 @@ func (oc *OrphanController) isResponsibleFor(orphan *longhorn.Orphan) bool {
 
 	defer func() {
 		if err == nil {
 			return
 		}
 
-		err = errors.Wrapf(err, "failed to delete orphan %v data", orphan.Name)
+		err = errors.Wrapf(err, "failed to delete %v orphan %v", orphan.Spec.Type, orphan.Name)
 		orphan.Status.Conditions = types.SetCondition(orphan.Status.Conditions,
 			longhorn.OrphanConditionTypeError, longhorn.ConditionStatusTrue, "", err.Error())
 	}()
 
 	// Make sure if the orphan nodeID and controller ID are same.
 	// If NO, just delete the orphan resource object and don't touch the data.
 	if orphan.Spec.NodeID != oc.controllerID {
-		log.Infof("Orphan nodeID %v is different from controllerID %v, so just delete the orphan resource object",
-			orphan.Name, oc.controllerID)
-		return nil
+		log.WithFields(logrus.Fields{
+			"orphanType": orphan.Spec.Type,
+			"orphanName": orphan.Name,
+			"orphanNode": orphan.Spec.NodeID,
+		}).Infof("Orphan does not belong to this controller. Skipping resource cleanup")
+		return true, nil
 	}
 
-	if types.GetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeDataCleanable).Status !=
-		longhorn.ConditionStatusTrue {
-		log.Infof("Only delete orphan %v resource object and do not delete the orphaned data", orphan.Name)
-		return nil
+	switch orphan.Spec.Type {
+	case longhorn.OrphanTypeEngineInstance:
+		isCleanupComplete, err = oc.cleanupOrphanedEngineInstance(orphan)
+	case longhorn.OrphanTypeReplicaInstance:
+		isCleanupComplete, err = oc.cleanupOrphanedReplicaInstance(orphan)
+	case longhorn.OrphanTypeReplicaData:
+		if types.GetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeDataCleanable).Status !=
+			longhorn.ConditionStatusTrue {
+			log.Infof("Only delete orphan %v resource object and do not delete the orphaned data store", orphan.Name)
+			return true, nil
+		}
+		err = oc.deleteOrphanedReplicaDataStore(orphan)
+		if err == nil || (err != nil && apierrors.IsNotFound(err)) {
+			isCleanupComplete = true
+			err = nil
+		}
+	default:
+		return false, fmt.Errorf("unknown orphan type %v to clean up orphaned resource for %v", orphan.Spec.Type, orphan.Name)
 	}
 
-	switch orphan.Spec.Type {
-	case longhorn.OrphanTypeReplica:
-		err = oc.deleteOrphanedReplica(orphan)
+	return isCleanupComplete, err
+}
+
+func (oc *OrphanController) cleanupOrphanedEngineInstance(orphan *longhorn.Orphan) (isCleanupComplete bool, err error) {
+	defer func() {
+		if err != nil {
+			err = errors.Wrapf(err, "cannot cleanup orphan engine instance %v", orphan.Name)
+		}
+	}()
+
+	instance, imName, err := oc.extractOrphanedInstanceInfo(orphan)
+	if err != nil {
+		return false, err
+	}
+
+	var spec *longhorn.InstanceSpec
+	if engineCR, err := oc.ds.GetEngineRO(instance); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return false, err
+		}
+		spec = nil
+	} else {
+		spec = &engineCR.Spec.InstanceSpec
+	}
+	return oc.cleanupOrphanedInstance(orphan, instance, imName, longhorn.InstanceManagerTypeEngine, orphan.Spec.DataEngine, spec)
```

Comment:
`orphan` is one of the parameters. Should we need `orphan.Spec.DataEngine` here?